### PR TITLE
Initial version of metadata plugin

### PR DESCRIPTION
**Pull request by jaredoconnell**
## Changes introduced with this PR

These changes make the plugin successfully gather ansible facts and output them.

Currently depends on main branch of python SDK due to Any type support requirement. Will change once it's released.

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: metadata_plugin.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    """
+    This is the output data structure for the success case.
+    """
+    metadata: typing.Any = field()
+
+
+@dataclass
+class ErrorOutput:
+    """
+    This is the output data structure in the error  case.
```
Comment (portante) [inline metadata_plugin.py:36]: If you come back to edit this file again, perhaps remove the extra space if you think of it.

File: metadata_plugin.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    """
+
```
Comment (portante) [inline metadata_plugin.py:38]: ... and the extra blank line.

File: metadata_plugin.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        new_list = []
+        for i in ansible_value:
+            new_list.append( convert_to_supported_type(i))
+        return new_list
+    if type_of_val == dict:
+        result = {}
+        for k in ansible_value:
+            result[ convert_to_supported_type(k)] =  convert_to_supported_type(ansible_value[k])
+        return result
+    if type_of_val == float or type_of_val == int or type_of_val == str or \
```
Comment (portante) [inline metadata_plugin.py:120]: Consider:

```
if type_of_val in (float, int, str, bool, type(None)):
```

File: metadata_plugin.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    type_of_val = type(ansible_value)
+    if type_of_val == list:
+        new_list = []
+        for i in ansible_value:
+            new_list.append( convert_to_supported_type(i))
+        return new_list
+    if type_of_val == dict:
+        result = {}
+        for k in ansible_value:
+            result[ convert_to_supported_type(k)] =  convert_to_supported_type(ansible_value[k])
```
Comment (portante) [inline metadata_plugin.py:118]: Extra spaces in this like throw off the code a bit.  If you think of it, and come back to this file for something else, perhaps remove them?

File: metadata_plugin.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    output =  convert_to_supported_type(host_ansible_facts)
+
+    return "success", SuccessOutput(output)
+
+def  convert_to_supported_type(ansible_value):
+    type_of_val = type(ansible_value)
+    if type_of_val == list:
+        new_list = []
+        for i in ansible_value:
+            new_list.append( convert_to_supported_type(i))
```
Comment (portante) [inline metadata_plugin.py:113]: Are we using `black` yet?

File: metadata_plugin.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        return new_list
+    if type_of_val == dict:
+        result = {}
+        for k in ansible_value:
+            result[ convert_to_supported_type(k)] =  convert_to_supported_type(ansible_value[k])
+        return result
+    if type_of_val == float or type_of_val == int or type_of_val == str or \
+        type_of_val == bool or type_of_val == type(None):
+        return ansible_value
+    elif type_of_val == AnsibleUnsafeText:
```
Comment (portante) [inline metadata_plugin.py:123]: Consider a small change to lines 123 - 127 as:

```
else:
    value = str(ansible_value)
    if type_of_val != AnsibleUnsafeText:
        print(f"Unknown type '{type_of_val}' with value '{value}'")
    return value
```

File: metadata_plugin.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+    try:
+        result = tqm.run(play)
+    finally:
+        tqm.cleanup()
+        if loader:
+            loader.cleanup_all_tmp_files()
+    
+    vars = variable_manager.get_vars(play=play, include_hostvars=True)
+    host_ansible_facts = vars["hostvars"][host]["ansible_facts"]
```
Comment (portante) [inline metadata_plugin.py:102]: Is it worth catching `KeyError` here in order to emit a more sensible error message when Ansible fails to return data in the format expected?

File: metadata_plugin.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    # Convert to dict
```
Comment (portante) [inline metadata_plugin.py:103]: Perhaps consider, "Convert to supported types"?

File: metadata_plugin.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    output =  convert_to_supported_type(host_ansible_facts)
```
Comment (portante) [inline metadata_plugin.py:104]: If you happen to come back in here, consider fixing the extra space ... or run `black` on the file.

File: metadata_plugin.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    return "success", SuccessOutput(output)
+
+def  convert_to_supported_type(ansible_value):
```
Comment (portante) [inline metadata_plugin.py:108]: Consider creating a function documentation string, and the extra whitespace.

**Review by portante — CHANGES_REQUESTED** [review]
Is the `output.txt` file necessary?

**Review by mfleader — APPROVED** [review]

**Review by jdowni000 — APPROVED** [review]

**Pull request merged**